drivers/bluetooth/hci/hci_da1469x.c

```diff
@@ -0,0 +1,495 @@
+		LOG_ERR("Failed to read UART (err %d)", read);
+		return;
+	}
+
+	net_buf_add(rx.buf, read);
+	rx.remaining -= read;
+
+	LOG_DBG("got %d bytes, remaining %u", read, rx.remaining);
+	LOG_DBG("Payload (len %u): %s", rx.buf->len, bt_hex(rx.buf->data, rx.buf->len));
+
+	if (rx.remaining) {
+		return;
+	}
+
+	buf = rx.buf;
+	rx.buf = NULL;
+
+	if (rx.type == H4_EVT) {
+		bt_buf_set_type(buf, BT_BUF_EVT);
+	} else {
+		bt_buf_set_type(buf, BT_BUF_ACL_IN);
+	}
+
+	reset_rx();
+
+	LOG_DBG("Putting buf %p to rx fifo", buf);
+	net_buf_put(&rx.fifo, buf);
+}
+
+static inline void read_header(void)
+{
+	switch (rx.type) {
+	case H4_NONE:
+		h4_get_type();
+		return;
+	case H4_EVT:
+		get_evt_hdr();
+		break;
+	case H4_ACL:
+		get_acl_hdr();
+		break;
+	case H4_ISO:
+		if (IS_ENABLED(CONFIG_BT_ISO)) {
+			get_iso_hdr();
+			break;
+		}
+		__fallthrough;
+	default:
+		CODE_UNREACHABLE;
+		return;
+	}
+
+	if (rx.have_hdr && rx.buf) {
+		if (rx.remaining > net_buf_tailroom(rx.buf)) {
+			LOG_ERR("Not enough space in buffer");
+			rx.discard = rx.remaining;
+			reset_rx();
+		} else {
+			copy_hdr(rx.buf);
+		}
+	}
+}
+
+static inline void process_rx(void)
+{
+	LOG_DBG("remaining %u discard %u have_hdr %u rx.buf %p len %u", rx.remaining, rx.discard,
+		rx.have_hdr, rx.buf, rx.buf ? rx.buf->len : 0);
+
+	if (rx.discard) {
+		rx.discard -= h4_discard(rx.discard);
+		return;
+	}
+
+	if (rx.have_hdr) {
+		read_payload();
+	} else {
+		read_header();
+	}
+}
+
+void cmac_read_req(void)
+{
+	while (cmac_mbox_has_data()) {
+		process_rx();
+	}
+}
+
+void cmac_rng_req(void)
+{
+	k_sem_give(&rng_sem);
+}
+
+static void rng_thread(void *p1, void *p2, void *p3)
+{
+	uint32_t word;
+
+	ARG_UNUSED(p1);
+	ARG_UNUSED(p2);
+	ARG_UNUSED(p3);
+
+	while (1) {
+		k_sem_take(&rng_sem, K_FOREVER);
+
+		while (cmac_rand_needs_data()) {
+			word = sys_rand32_get();
+			cmac_rand_fill(&word, 1);
+		}
+
+		cmac_signal();
+	}
+}
+
+static int bt_da1469x_open(void)
+{
+	k_tid_t tid;
+
+	tid = k_thread_create(&rx_thread_data, rx_thread_stack,
+			      K_KERNEL_STACK_SIZEOF(rx_thread_stack),
+			      rx_thread, NULL, NULL, NULL,
+			      K_PRIO_COOP(CONFIG_BT_RX_PRIO),
+			      0, K_NO_WAIT);
+	k_thread_name_set(tid, "bt_rx_thread");
+
+	k_sem_init(&rng_sem, 0, 1);
+
+	tid = k_thread_create(&rng_thread_data, rng_thread_stack,
+			      K_KERNEL_STACK_SIZEOF(rng_thread_stack),
+			      rng_thread, NULL, NULL, NULL,
+			      K_PRIO_COOP(CONFIG_BT_RX_PRIO),
+			      0, K_NO_WAIT);
+	k_thread_name_set(tid, "bt_rng_thread");
+
+	cmac_enable();
+	irq_enable(CMAC2SYS_IRQn);
+
+	return 0;
+}
+
+#ifdef CONFIG_BT_HCI_HOST
+static int bt_da1469x_close(void)
+{
+	irq_disable(CMAC2SYS_IRQn);
+	cmac_disable();
+
+	return 0;
+}
+#endif /* CONFIG_BT_HCI_HOST */
+
+static int bt_da1469x_send(struct net_buf *buf)
+{
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_ACL_OUT:
+		LOG_DBG("ACL: buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+		net_buf_push_u8(buf, H4_ACL);
```

Comment:
To be able to safely do this, the driver first needs to make sure it requests sufficient headroom in the buffers:
https://github.com/zephyrproject-rtos/zephyr/blob/eddb3822d2aa746026837252b3e337b0c4ef067a/subsys/bluetooth/host/Kconfig#L75-L93

I know this is not the prettiest solution, since that list of defaults keeps growing. Once the HCI driver interface has been converted to a proper Zephyr driver API (see #51503), this information could potentially come from a devicetree property.

That said, how much have you actually tested this driver? Without requesting the buffer headroom you should be getting a buffer underflow and potential memory corruption.

Reply:
> Seems like BT_HCI_RESERVE was removed in 3.6 and headroom is now always reserved.

Yes, you can disregard this comment thanks to the removal.